src/cljx/schema/spec/collection.cljx

```diff
@@ -0,0 +1,85 @@
+(ns schema.spec.collection
+  "A collection spec represents a collection of elements,
+   each of which is itself schematized."
+  (:require
+   [schema.utils :as utils]
+   [schema.spec.core :as spec]))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Collection Specs
+
+(defn- element-transformer [e params then]
+  (let [parser (:parser e)
+        c (spec/sub-checker e params)]
+    #+clj (fn [^java.util.List res x]
+            (then res (parser (fn [t] (.add res (if (utils/error? t) t (c t)))) x)))
+    #+cljs (fn [res x]
+             (then res (parser (fn [t] (swap! res conj (if (utils/error? t) t (c t)))) x)))))
+
+#+clj ;; for performance
+(defn- has-error? [^java.util.List l]
+  (let [it (.iterator l)]
+    (loop []
+      (if (.hasNext it)
+        (if (utils/error? (.next it))
+          true
+          (recur))
+        false))))
+
+#+cljs
+(defn- has-error? [l]
+  (some utils/error? l))
+
+(defrecord CollectionSpec [pre constructor elements on-error]
+  spec/CoreSpec
+  (subschemas [this] (map :schema elements))
+  (checker [this params]
+    (let [constructor (if (:return-walked? params) constructor (fn [_] nil))
+          t (reduce
+             (fn [f e]
+               (element-transformer e params f))
+             (fn [_ x] x)
+             (reverse elements))]
+      (fn [x]
+        (or (pre x)
+            (let [res #+clj (java.util.ArrayList.) #+cljs (atom [])
+                  remaining (t res x)
+                  res #+clj res #+cljs @res]
+              (if (or (seq remaining) (has-error? res))
+                (utils/error (on-error x res remaining))
+                (constructor res))))))))
+
+(defn collection-spec
+  "A collection represents a collection of elements, each of which is itself
+   schematized.  At the top level, the collection has a precondition
+   (presumably on the overall type), a constructor for the collection from a
+   sequence of items, an element spec, and a function that constructs a
+   descriptive error on failure.
+
+   The element spec is a sequence of maps, each of which provides an element
+   schema, cardinality, parser (allowing for efficient processing of
+   structured collections), and optional error wrapper."
+  [pre ;- spec/Precondition
+   constructor ;- (s/=> s/Any [(s/named s/Any 'checked-value)])
```

Comment:
oh! :sheep: :sheep: why not derive the constructor from the schema? i.e. `#(into (empty s) %)`?

i know that this is less flexible, but in all real-world cases (vector, map, set) this works as the constructor. i guess it would be cleaner to have it be taken from the data value being checked instead of the schema..

also, it would support preserving the type of sequence (list, vector) and the type of map (i.e. sorted vs hashmap)?